Feature/2019 08 30 nick van hoof monitoring serverless apps on aws (#555)
ordina-jworks/ordina-jworks.github.io: _posts/2019-10-15-Monitoring-serverless-apps-on-AWS.md

category: Cloud
comments: false
---

# Table of content
* [What about](#what-about)

* [Challenges of Serverless applications](#challenges-of-serverless-applications)
* [Challenge 1: Finding the error in a distributed serverless landscape](#challenge-1-finding-the-error-in-a-distributed-serverless-landscape)
* [Solution 1: Structured logging](#solution-1-structured-logging)

* [Challenge 2: Finding performance bottlenecks](#challenge-2-finding-performance-bottlenecks)
* [Solution 2: Distributed tracing with Xray](#solution-2-distributed-tracing-with-aws-xray)

* [Challenge 3: Testing whether our application still behaves as expected](#challenge-3-testing-whether-our-application-still-behaves-as-expected)
* [Solution 3a: Smoke Testing](#solution-3a-smoke-testing)
* [Solution 3b: Load Testing](#solution-3b-load-testing)

* [Side note on CloudWatch Dashboards](#side-note-on-cloudwatch-dashboards)
* [Conclusion](#conclusion)
* [Resources](#resources)

# What about?
Serverless is a great technology that comes with the advantage of being scalable, durable and high available.  
It allows you to decouple functionality into multiple serverless Functions.  

> Of course with new technologies come also new challenges.

Reviewer: ```suggestion
But with new technologies comes new challenges.
```

Having an application that exist of a lot of decoupled lambda functions means that your serverless landscape will be heavily distributed.  

> I mean that a lot of stuff happens in a lot of different places.  

Reviewer: ```suggestion
I mean that's a lot of stuff happening in a lot of different places.  
```

This means that a distributed serverless landscape has to be observable.  
Let's see some of the best practices on how to make your serverless landscape observable. 

# Challenges of Serverless applications
What does a typical serverless application look like?  

Let's look at an app that was build for a conference. 
Speakers can create a session that they want to speak about. 
People can also retrieve all sessions that have already been submitted.  
When a new session is created a slack notification is sent out.

<div style="text-align: center;">
  <img src="/img/2019-10-15-Monitoring-serverless-apps-on-AWS/simpler-architecture.png" width="100%" height="100%">
</div>

We can identify certain milestones that indicate that a request has passed this milestone.

<div style="text-align: center;">
  <img src="/img/2019-10-15-Monitoring-serverless-apps-on-AWS/simpler-architecture-with-milestones.png" width="100%" height="100%">
</div>

The serverless architecture above is actually quite small.
I've seen architectures containing tens of Lambda Functions and other AWS services.

Some of the challenges that come with a serverless architecture are:

> 1. It might wrong somewhere in my distributed landscape.

Reviewer: ```suggestion
1. It might crash somewhere in my distributed landscape.
```

</div>

# Challenge 2: Finding performance bottlenecks
I wrote a `Logs Insights` query that allows me to check how long it took for a request to pass through the whole landscape.
That means from the moment the creation request arrived till the moment we send out a slack notification for it.

```bash
fields @timestamp, @message
| filter  milestone="CREATE_REQUEST_RECEIVED" or milestone="SLACK_NOTIFICATION_NEW_SESSION_SENT"
| stats (latest(@timestamp) - earliest(@timestamp))/1000 as LeadTimeInSeconds by traceId
| filter LeadTimeInSeconds!=0
| sort LeadTimeInSeconds desc
| limit 20
```
    
<div style="text-align: center;">
  <img src="/img/2019-10-15-Monitoring-serverless-apps-on-AWS/slack-notification-lead-time.png" width="100%" height="100%">
</div>

We see that even when the system is warm, it takes us up to 10 seconds to send out a slack notification.
We need to dig into the performance of our lambda functions using `AWS Xray`.

## Solution 2: distributed tracing with AWS Xray

<div style="text-align: center;">
  <img src="/img/2019-10-15-Monitoring-serverless-apps-on-AWS/xray.png" width="10%" height="10%">
</div>

> Xray helps us understand the behaviour of our system and thus allows us to analyze the performance of specific parts.

Reviewer: ```suggestion
Xray helps us understand the behavior of our system and thus allows us to analyze the performance of specific parts.
```

A trace is actually build up from multiple segments.

It does this by:
* Sampling your requests. 
By default Xray will trace 5% of your requests.
* Tracing calls made by ths AWS SDK.
This happens automatically when you use Xray as a dependency for your project.
* Creating custom segments.
You can create your own segments as you see below.

```java
Subsegment subsegment = AWSXRay.beginSubsegment("Sessions.saveSessionDynamoDB");

subsegment.putAnnotation("storeInDatabase", "DynamoDB");
subsegment.putMetadata("company", "Ordina");

mockingIssues(sessionDynamoDao);

repository.saveSession(sessionDynamoDao);

AWSXRay.endSubsegment();
```

Here is an example of the Xray service map.

<div style="text-align: center;">
  <img src="/img/2019-10-15-Monitoring-serverless-apps-on-AWS/xray-service-map.png" width="100%" height="100%">
</div>

Xray doesn't trace async requests (yet).
That means that publishing on an `SNS` topic or going via a `DynamoDB Stream` is not part of the full trace but will show up as a new client in the service map.
Recently tracing over `SQS` was added.

When we click the lambda service we can see the response distribution.

> This visualizes how quickly this lambda function responded.

Reviewer: ```suggestion
This visualizes how quickly the lambda function responded.
```

<div style="text-align: center;">
  <img src="/img/2019-10-15-Monitoring-serverless-apps-on-AWS/xray-response-distribution.png" width="100%" height="100%">
</div>

Something is definitely wrong here since even the quickest executions take more than 3 seconds.
We can dig deeper by clicking `view traces`.

<div style="text-align: center;">
  <img src="/img/2019-10-15-Monitoring-serverless-apps-on-AWS/xray-overview-traces.png" width="100%" height="100%">
</div> 

Digging even deeper into one of these traces we can see how long every segment of this trace took.

<div style="text-align: center;">
  <img src="/img/2019-10-15-Monitoring-serverless-apps-on-AWS/xray-segments-of-trace.png" width="100%" height="100%">
</div> 

Below we see that first some data was saved to the `caching table`.
This happened blazingly quick in 8.0 ms.  

We see however that the `Sessions.saveSessionDynamoDB` segment took over 3.0 seconds.

> Of these 3 seconds only 8 ms where spend actually saving the request.

Reviewer: ```suggestion
Of these 3 seconds, only 8 ms was spent actually saving the request.
```

We found our performance bottleneck.
Something is waiting around in the `Sessions.saveSessionDynamoDB` segment.
In this case it was me introducing an artificial `Thread.sleep()`.

Hooray, we found the performance bottleneck.  
Challenge 2 completed.

<div style="text-align: center;">
  <img src="/img/2019-10-15-Monitoring-serverless-apps-on-AWS/completed.png" width="10%" height="10%">
</div>

# Challenge 3: Testing whether our application still behaves as expected
We can't run our complete cloud infrastructure on our local machine.  

> So when we change an redeploy, we should test if our system is still behaving as it should.

Reviewer: ```suggestion
So when we change and redeploy, we should test if our system is still behaving as it should.
```

</div>

I create this test using `JUnit` and mocked the http endpoints with [wiremock](http://wiremock.org/){:target="_blank"}.
`Wiremock` is a great tool to mock http endpoints that I personally use a lot.  
You can ask Wiremock to create certain endpoints and configure the response for it.  
Below you see me creating the `/sessions/forward` endpoint.

```bash
curl -X POST \
  $wiremock_url \
  -H 'Content-Type: application/json' \
  -d '{
    "request": {
        "method": "POST",
        "url": "/session/forward"
    },
    "response": {
        "status": 200,
        "body": "I have received the session correctly",
         "delayDistribution": {
                    "type": "lognormal",
                    "median": 100,
                    "sigma": 0.1
           },
        "headers": {
            "Content-Type": "text/plain"
        }
    }
}'
```

## Solution 3b: load testing
Yes, serverless scales automatically.
But things might not always behave as expected.
Listening on events of a `Kinesis` stream for example is only possible with one Lambda function per `Shard`.
Thus limiting your throughput if you don't watch out.

> To run my loaddtest I use [artillery](https://artillery.io){:target="_blank"}. 

Reviewer: ```suggestion
To run my load test I use [artillery](https://artillery.io){:target="_blank"}. 
```

AWS allows you to bundle the results of these queries into dashboards via `CloudWatch Dashboards`.

Below you see how I made a dashboard that visualizes the number of invocations and associated costs per lambda function.

<div style="text-align: center;">
  <img src="/img/2019-10-15-Monitoring-serverless-apps-on-AWS/billing-overview.png" width="100%" height="100%">
</div>

To find the total cost I used the following query:
```bash
filter @type = "REPORT"
| fields @memorySize/1000000 as MemorySetInMB, @billedDuration/1000*MemorySetInMB/1024 as BilledDurationInGBSeconds, @logStream
| stats sum(BilledDurationInGBSeconds) as TotalBilledDurationInGBSeconds, sum(BilledDurationInGBSeconds) * 0.00001667 as TotalCostInDollar
```

To get the stats per Lambda function I did:
```bash
filter @type="REPORT"
| fields @memorySize/1000000 as MemorySetInMB, @billedDuration/1000*MemorySetInMB/1024 as BilledDurationInGBSeconds
| stats 
count(@billedDuration) as NumberOfInvocations,
ceil(avg(@duration)) as AverageExecutionTime,
max(@duration) as MaxExecutionTime,
sum(BilledDurationInGBSeconds) * 0.00001667 as TotalCostInDollar
```

# Third party tools

> We just saw a the things that we can do to increase the observability of our serverless landscape.  

Reviewer: ```suggestion
We just saw the things that we can do to increase the observability of our serverless landscape.  
```

> To achieve this you'll have to do some custom work:

Reviewer: ```suggestion
To achieve this, you'll have to do some custom work:
```

* setup structured logging
* pass on a traceId
* create dashboards in CloudWatch
* get familiar with the `Logs Insights` query language
* configure the right alarms in CloudWatch
* setup the infrastructure to notify you when an error appears in your landscape

These are things that take time.  
And time is money.  

You can also consider using this money to work with a third party tool.
This tool then allows you to monitor and troubleshoot your serverless applications.

Personally I have used [Lumigo](https://lumigo.io){:target="_blank"}  to achieve just that.  

> Many of the things that you have to custom build will be provided out of the box.

Reviewer: ```suggestion
The things you have to customize, will be provided out of the box.
```

> You get an high level dashboard to visualize the problems in your application.

Reviewer: ```suggestion
You get a high-level dashboard to visualize the problems in your application.
```

<div style="text-align: center;">
  <img src="/img/2019-10-15-Monitoring-serverless-apps-on-AWS/lumigo-dashboard.png" width="80%" height="80%">
</div>

It will also automatically trace a request through your landscape by creating a transaction.
This is about the same as we did by forwarding traceIds. 
 
It comes with a handy visualization.
On the left you see the transaction while on the right it gives you the logs that correspond with the error that appeared.

<div style="text-align: center;">
  <img src="/img/2019-10-15-Monitoring-serverless-apps-on-AWS/lumigo-transaction-and-error.png" width="80%" height="80%">
</div>

This allows you to drill down to the source of the error quickly.  
On top of that you can also create the necessary alerts to notify you in case things go south.

These tools often have a [free tier](https://platform.lumigo.io/signup){:target="_blank"}  that allows you to explore the product.

# Conclusion

We improved the observability of our system by implementing structured logs and using appropriate testing and tooling.

> By doing this it becomes way easier to monitor your system and create visibility on it's behavior.

Reviewer: ```suggestion
By doing this, it becomes way easier to monitor your system and create visibility on its behavior.
```

Remember that:

> * you need structured logging to get the maximum out of your logs.

Reviewer: do not end with period in a listing
